hermes-client/src/main/java/pl/allegro/tech/hermes/client/ReactiveHermesClient.java

```diff
@@ -0,0 +1,324 @@
+                        return Result.failure(new ShouldRetryException(response), attempt);
+                    } else {
+                        return Result.attempt(response, attempt);
+                    }
+                });
+    }
+
+    private Retry prepareRetrySpec(HermesMessage message) {
+        if (!retrySleep.isZero()) {
+            return Retry.max(maxRetries)
+                    .doAfterRetry(signal -> handleFailedAttempt(message, signal.totalRetries() + 1));
+        } else {
+            return Retry.backoff(maxRetries, retrySleep)
+                    .maxBackoff(maxRetrySleep)
+                    .doAfterRetry(signal -> handleFailedAttempt(message, signal.totalRetries() + 1));
+        }
+    }
+
+    private Mono<Integer> getNextAttempt() {
+        return Mono.subscriberContext()
+                .map(ctx -> ctx.getOrDefault(RETRY_CONTEXT_KEY, HermesRetryContext.emptyRetryContext())
+                        .getAndIncrementAttempt()
+                );
+    }
+
+    private Mono<HermesResponse> sendOnce(HermesMessage message) {
+        return Mono.defer(() -> {
+                    long startTime = System.nanoTime();
+                    try {
+                        return sender.sendReactively(URI.create(uri + message.getTopic()), message)
+                                .onErrorResume(e -> Mono.just(HermesResponseBuilder.hermesFailureResponse(e, message)))
+                                .doOnNext(resp -> {
+                                    long latency = System.nanoTime() - startTime;
+                                    messageDeliveryListeners.forEach(l -> l.onSend(resp, latency));
+                                });
+                    } catch (Exception e) {
+                        return Mono.error(e);
+                    }
+                }
+        );
+    }
+
+    private Mono<HermesResponse> completedWithShutdownException() {
+        return Mono.error(new HermesClientShutdownException());
+    }
+
+    public Mono<Void> closeAsync(long pollInterval) {
+        shutdown = true;
+        CompletableFuture<Void> voidCompletableFuture = new HermesClientTermination(pollInterval)
+                .observe(() -> currentlySending.get() == 0)
+                .whenComplete((response, ex) -> scheduler.dispose());
+        return Mono.fromFuture(voidCompletableFuture);
+    }
+
+    public void close(long pollInterval, long timeout) {
+        closeAsync(pollInterval).block(Duration.ofMillis(timeout));
+    }
+
+    private void handleMaxRetriesExceeded(HermesMessage message, int attemptCount) {
+        messageDeliveryListeners.forEach(l -> l.onMaxRetriesExceeded(message, attemptCount));
+        logger.error("Failed to send message to topic {} after {} attempts",
+                message.getTopic(), attemptCount);
+    }
+
+    private void handleFailedAttempt(HermesMessage message, long attemptCount) {
+        messageDeliveryListeners.forEach(l -> l.onFailedRetry(message, (int) attemptCount));
```

Comment:
I think we should be careful and enforce a difference between "retry" and "attempt".

Currently we have a problem with metrics in non-reactive client with distinguishing retry counter from total request counter. As a consequence, one cannot be sure, for example, whether the "retries.success" indicated that the retry was performed or not.

It is discussable whether we want to intentionally preserve that inconsistency and then fix it both here and in the non-reactive client for the sake of coherence between the reactive and non-reactive-client; or do we want to fix it in reactive client right now.